frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -3956,6 +4011,145 @@ bool isTypeDefaultInitializable(Context* context, const Type* t) {
+        // sync/single implicit reads are removed. 12/8/23
+        fromConst = true;
+      } else {
+        // In general, try to resolve the type's 'init='/'=', and examine it to
+        // determine copy/assignability, respectively.
+        const TypedFnSignature* testResolvedSig =
+            (checkCopyable ? tryResolveInitEq(context, ast, ct, ct)
+                           : tryResolveEq(context, ast, ct));
+        if (testResolvedSig) {
+          if (testResolvedSig->untyped()->isCompilerGenerated()) {
+            // Check for owned class fields.
+
+            bool containsNonNilableOwned = false;
+            bool containsNilableOwned = false;
+            auto resolvedFields =
+                fieldsForTypeDecl(context, ct, DefaultsPolicy::USE_DEFAULTS);
+            for (int i = 0; i < resolvedFields.numFields(); i++) {
+              auto fieldType = resolvedFields.fieldType(i).type();
+              if (auto classTy = fieldType->toClassType()) {
+                if (classTy->decorator().isManaged() &&
+                    classTy->manager()->isAnyOwnedType()) {
+                  if (classTy->decorator().isNonNilable()) {
+                    containsNonNilableOwned = true;
+                  } else {
+                    containsNilableOwned = true;
+                  }
+                }
+              }
+            }
+
+            fromRef = !containsNonNilableOwned && containsNilableOwned;
+            fromConst = !containsNonNilableOwned && !containsNilableOwned;
+          } else {
+            // Check intent of formal to copy/assign from.
+
+            // For init=, formals are (this, other). For =, formals are (lhs,
+            // rhs), unless it is a method in which case they are (this, lhs,
+            // rhs). Get the index of the 'other' or 'rhs' formal.
+            int otherFormalNum = 1;
+            if (!checkCopyable && testResolvedSig->untyped()->isMethod()) {
+              otherFormalNum = 2;
+            }
+            CHPL_ASSERT(testResolvedSig->numFormals() == (otherFormalNum + 1) &&
+                        "unexpected formals");
+            auto otherTy = testResolvedSig->formalType(otherFormalNum);
+            CHPL_ASSERT(!otherTy.isNonConcreteIntent() &&
+                        "should have resolved concrete intent by now");
+            auto otherIntent = otherTy.kind();
+
+            fromConst = (otherIntent == QualifiedType::IN ||
+                         otherIntent == QualifiedType::CONST_IN ||
+                         otherIntent == QualifiedType::CONST_REF);
+            fromRef = !fromConst;
+          }
+        }
+      }
+    }
+  }
+
+  std::pair<bool, bool> result = {fromConst, fromRef};
+  return QUERY_END(result);
+}
+
+void getCopyOrAssignableInfo(Context* context, const Type* t,
+                                    bool* fromConst, bool* fromRef,
+                                    bool checkCopyable) {
+  if (t->isCompositeType() || t->isClassType()) {
```

Comment:
Perhaps this is better written as `if (t->getCompositeType())`

Reply:
I did have that originally, but then discovered `ClassType` is not a `CompositeType`, so it didn't have the right behavior

Reply:
Oh, I see your below comment suggesting taking the logic for `ClassType`s out of the query -- I'll do that

Reply:
`t->getCompositeType()` will return non-null for ClassType and RecordType etc. It's different from `isCompositeType`. But as you said that might well be moot here.

Reply:
Oh, I had just read that as `toCompositeType`, that's good to know about.